test/experimental/features/new_trigger_filters/filters.go

```diff
@@ -0,0 +1,123 @@
+	. "knative.dev/reconciler-test/pkg/eventshub/assert"
+	"knative.dev/reconciler-test/pkg/feature"
+	"knative.dev/reconciler-test/pkg/manifest"
+	"knative.dev/reconciler-test/resources/svc"
+
+	"knative.dev/eventing/test/rekt/resources/broker"
+	"knative.dev/eventing/test/rekt/resources/trigger"
+)
+
+// FiltersFeatureSet creates a feature set for testing the broker implementation of the new trigger filters experimental feature
+// (aka Cloud Events Subscriptions API filters). It requires a created and ready Broker resource with brokerName.
+//
+// The feature set tests four filter dialects: exact, prefix, suffix and sql (aka CloudEvents SQL).
+func FiltersFeatureSet(brokerName string) *feature.FeatureSet {
+	matchedEvent := FullEvent()
+	unmatchedEvent := MinEvent()
+	unmatchedEvent.SetType("com.wrong.type")
+	unmatchedEvent.SetSource("com.wrong.source")
+
+	var features []feature.Feature
+	tests := map[string]struct {
+		filters string
+		step    feature.StepFn
+	}{
+		"Exact filter": {
+			filters: fmt.Sprintf(snippetFor("exact"), matchedEvent.Type(), matchedEvent.Source()),
+		},
+		"Prefix filter": {
+			filters: fmt.Sprintf(snippetFor("prefix"), matchedEvent.Type()[:4], matchedEvent.Source()[:4]),
+		},
+		"Suffix filter": {
+			filters: fmt.Sprintf(snippetFor("suffix"), matchedEvent.Type()[5:], matchedEvent.Source()[5:]),
+		},
+		"CloudEvents SQL filter": {
+			filters: fmt.Sprintf(`- sql: "type = '%s' AND source = '%s'" `, matchedEvent.Type(), matchedEvent.Source()),
+		},
+	}
+
+	for name, fs := range tests {
+		matchedSender := feature.MakeRandomK8sName("sender")
+		unmatchedSender := feature.MakeRandomK8sName("sender")
+		subscriber := feature.MakeRandomK8sName("subscriber")
+		triggerName := feature.MakeRandomK8sName("viaTrigger")
+
+		f := feature.NewFeatureNamed(name)
+
+		f.Setup("Install trigger subscriber", eventshub.Install(subscriber, eventshub.StartReceiver))
+
+		// Set the Trigger subscriber.
+		cfg := []manifest.CfgFn{
+			trigger.WithSubscriber(svc.AsKReference(subscriber), ""),
+			WithNewFilters(fs.filters),
+		}
+
+		f.Setup("Install trigger", trigger.Install(triggerName, brokerName, cfg...))
+		f.Setup("Wait for trigger to become ready", trigger.IsReady(triggerName))
+
+		f.Setup("Install matched event sender", func(ctx context.Context, t feature.T) {
+			u, err := broker.Address(ctx, brokerName)
+			if err != nil || u == nil {
+				t.Error("failed to get the address of the broker", brokerName, err)
+			}
+			eventshub.Install(matchedSender, eventshub.StartSenderURL(u.String()), eventshub.InputEvent(matchedEvent))(ctx, t)
+		})
+
+		f.Setup("Install unmatched event sender", func(ctx context.Context, t feature.T) {
+			u, err := broker.Address(ctx, brokerName)
+			if err != nil || u == nil {
+				t.Error("failed to get the address of the broker", brokerName, err)
+			}
+			eventshub.Install(unmatchedSender, eventshub.StartSenderURL(u.String()), eventshub.InputEvent(unmatchedEvent))(ctx, t)
+		})
```

Comment:
Since the spec specify as must that only events that are sent after a trigger is ready must be sent to the subscriber, this might be racy if the sender starts before the trigger is ready, so I'd move `install ... event sender` to the requirement phase.
```suggestion
		f.Setup("Wait for trigger to become ready", trigger.IsReady(triggerName))

		f.Requirement("Install matched event sender", func(ctx context.Context, t feature.T) {
			u, err := broker.Address(ctx, brokerName)
			if err != nil || u == nil {
				t.Error("failed to get the address of the broker", brokerName, err)
			}
			eventshub.Install(matchedSender, eventshub.StartSenderURL(u.String()), eventshub.InputEvent(matchedEvent))(ctx, t)
		})

		f.Requirement("Install unmatched event sender", func(ctx context.Context, t feature.T) {
			u, err := broker.Address(ctx, brokerName)
			if err != nil || u == nil {
				t.Error("failed to get the address of the broker", brokerName, err)
			}
			eventshub.Install(unmatchedSender, eventshub.StartSenderURL(u.String()), eventshub.InputEvent(unmatchedEvent))(ctx, t)
		})
```

Reply:
good catch 👍🏽